documentation/jetty-documentation/src/main/asciidoc/programming-guide/arch-threads.adoc

```diff
@@ -0,0 +1,210 @@
+
+Jetty's default thread pool implementation is link:{javadoc-url}/org/eclipse/jetty/util/thread/QueuedThreadPool.html[`QueuedThreadPool`]
+`QueuedThreadPool` integrates with the xref:pg-arch-bean[Jetty component model], implements `Executor`, provides a `TryExecutor` implementation (discussed in the xref:pg-arch-threads-execution-strategy-adaptive[adaptive execution strategy section]), and supports xref:pg-arch-threads-thread-pool-virtual-threads[virtual threads] (introduced as a preview feature in Java 19).
+
+`QueuedThreadPool` can be configured with a `maxThreads` value.
+
+However, some of the Jetty components (such as the xref:pg-arch-io-selector-manager[selectors]) permanently steal threads for their internal use, or rather `QueuedThreadPool` leases some threads to these components.
+These threads are reported by `QueuedThreadPool.leasedThreads` and are not available to run application code.
+
+`QueuedThreadPool` can be configured with a `reservedThreads` value.
+This value represents the maximum number of threads that can be reserved and used by the `TryExecutor` implementation.
+A negative value for `QueuedThreadPool.reservedThreads` means that the actual value will be heuristically derived from the number of CPU cores and `QueuedThreadPool.maxThreads`.
+A value of zero for `QueuedThreadPool.reservedThreads` means that reserved threads are disabled, and therefore the xref:pg-arch-threads-execution-strategy-epc[Execute-Produce-Consume mode] is never used -- the xref:pg-arch-threads-execution-strategy-pec[Produce-Execute-Consume mode] is always used instead.
+
+[[pg-arch-threads-thread-pool-virtual-threads]]
+===== Virtual Threads
+Virtual threads have been introduced in Java 19 as a preview feature.
+
+NOTE: In Java versions where virtual threads are a preview feature, remember to add `+--enable-preview+` to the command line options to use virtual threads.
+
+`QueuedThreadPool` can be configured to use virtual threads via:
+
+[source,java,indent=0]
+----
+include::{doc_code}/org/eclipse/jetty/docs/programming/QueuedThreadPoolDocs.java[tags=virtual-threads]
+----
+
+`AdaptiveExecutionStrategy` makes use of this setting when it determines that a task should be run with the xref:pg-arch-threads-execution-strategy-pec[`Produce-Execute-Consume` mode]: rather than submitting the task to `QueuedThreadPool` to be run in a platform thread, it runs the task in a new virtual thread.
+
+[NOTE]
+====
+Enabling virtual threads in `QueuedThreadPool` will default the number of reserved threads to zero, unless the number of reserved threads is explicitly configured to a positive value.
+
+Defaulting the number of reserved threads to zero ensures that the xref:pg-arch-threads-execution-strategy-pec[Produce-Execute-Consume mode] is always used, which means that virtual threads will always be used for blocking tasks.
+====
```

Comment:
A common question that comes up is how threading is used during an HTTP exchange, is this the document/place to discuss this?